source/common/secret/sds_api.cc

```diff
@@ -0,0 +1,82 @@
+#include "common/secret/sds_api.h"
+
+#include <unordered_map>
+
+#include "envoy/api/v2/auth/cert.pb.validate.h"
+
+#include "common/config/resources.h"
+#include "common/config/subscription_factory.h"
+#include "common/protobuf/utility.h"
+#include "common/ssl/tls_certificate_config_impl.h"
+
+namespace Envoy {
+namespace Secret {
+
+SdsApi::SdsApi(const LocalInfo::LocalInfo& local_info, Event::Dispatcher& dispatcher,
+               Runtime::RandomGenerator& random, Stats::Store& stats,
+               Upstream::ClusterManager& cluster_manager, Init::Manager& init_manager,
+               const envoy::api::v2::core::ConfigSource& sds_config, std::string sds_config_name)
+    : local_info_(local_info), dispatcher_(dispatcher), random_(random), stats_(stats),
+      cluster_manager_(cluster_manager), sds_config_(sds_config), sds_config_name_(sds_config_name),
+      secret_hash_(0) {
+  init_manager.registerTarget(*this);
```

Comment:
I feel this is a bit wrong from the rds fix etc, in secret manager if there is an existing SdsApi, we won't call constructor again, but does that means if two cluster shares same secret, the second one wouldn't get its init manager registering?

Reply:
We knew this issue.  e.g. if two listeners share same sds_api,  the second init_manager will not be used.   the second listener will not wait for sds_subscription, will be ready right away.  During such transit state, the requests to the second one will be rejected, but envoy will not crash.   This is acceptable situation without introducing more code changes. 
We can add a TODO for such issue.  I already have plan to fix it.  It will be called chained_init_manager.   Multiple init_manager can be chained together to behave as one init_manager.  This can be a separate PR.

Reply:
yes please add a TODO.